### PR TITLE
home-manager: do not build news when using flake

### PR DESCRIPTION
**Pull request by polykernel**
Currently, the `buildNews` and `doBuildAttrs` are always called
unconditionally even if a flake configuration is specified. This cause
it to always fail prior to the actual build performed by `doBuildAttrs`
because `setConfigFIle` can not find the home-manager configuration file.
As a result, an error message specifying no configuration file is shown.

Furthermore, if a user has remnant legacy configuration, the `doSwitch` and
`doBuild` functions will effectively build the activationPackage twice, with
the legacy configuration overriding the flake configuration.

A conditional check for FLAKE_CONFIG_URI was added to mitigate this by building
the legacy configuration when no flake configuration is present. There is one
exception which is when a flake configuration exists in the default location, where
the user can not build the legacy configuration as along as the file is present.
However, the tradeoff is acceptable as it matches current behavior when FLAKe_CONFIG_URI
is set for instantiation, and an user is unlikely to simulataneously switch
between the two mechanisms.

An abstract function for building flakes `doBuildFlake` was created to match
`doBuildAttrs` for  manageing options and build flags.

The --no-write-lock-file flag was removed from the --debug case as it is already
matched previously at the --recreate-lock-file case.

### Checklist

- [ ] Change is backwards compatible.

- [ ] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Review by berbiche — CHANGES_REQUESTED** [review]

File: home-manager/home-manager [hunk 1]
```diff
@@ -616,8 +627,7 @@ while [[ $# -gt 0 ]]; do
             shift
             ;;
         --debug|--keep-failed|--keep-going|--show-trace\
-            |--substitute|--no-substitute|--impure\
-            |--no-write-lock-file)
```
[the remark this answers is not in the document]
Comment (berbiche) [reply home-manager/home-manager:620]: Besides that, LGTM!

**Review by berbiche — APPROVED** [review]
I don't have a standalone HM configuration ATM so I trust that it works for you :smile:

**Comment by SuperSandro2000** [comment]
Does this fix the following?

```
home-manager news
error: infinite recursion encountered

       at $HOME/src/nixpkgs/lib/modules.nix:365:28:

          364|         builtins.addErrorContext (context name)
          365|           (args.${name} or config._module.args.${name})
             |                            ^
          366|       ) (lib.functionArgs f);
(use '--show-trace' to show detailed location information)
$HOME/.nix-profile/bin/home-manager: line 425: /tmp/home-manager-build.M3nPK4D340/news-info.sh: No such file or directory
```

Also how do I now get news on flakes?

**Comment by polykernel** [comment]
@SuperSandro2000  I don't think I changed anything relating to the presentNews and buildNews functions, I think the only same that could have change is news(which must be built using search paths) isn't built unconditionally if flake is used. The previous behavior was to build the flake activationPackage first, followed by legacy activationPackage then news. Unfortunately, this means it is currently impossible to build news with flakes, however I am working to integrate home-manager.nix with modules somehow so it can be exposed in the root flake.

**Comment by SuperSandro2000** [comment]
> was to build the flake activationPackage first, followed by legacy activationPackage then news.

I might have seen that before already. Always wondered why it takes so much longer.

> I am working on integrating home-manager.nix with modules.nix somehow so it can be exposed in the root flake.

If you remember please ping me for testing.

I think my above error was due to a bad overlay which is not obvious at all.

**Pull request merged**